File: python/232.py
"""Project Euler Problem 232: The Race.

Two players alternatively toss a coin. Find the probability player 2 wins
if she plays optimally.
"""

from __future__ import annotations

from typing import List


def pow_int(base: int, exp: int) -> int:
    """Return base^exp as integer."""
    return base**exp


def solve() -> float:
    """Solve Problem 232."""
    N = 100

    # dp[score1][score2] = probability player 2 wins
    dp: List[List[float]] = [[0.0] * (2 * N) for _ in range(N + 1)]

    # Base cases: player 2 wins if score2 >= N
    for score1 in range(N + 1):
        for score2 in range(N, 2 * N):
            dp[score1][score2] = 1.0

    # Fill DP table backwards
    for score1 in range(N - 1, -1, -1):
        for score2 in range(N - 1, -1, -1):
            for T in range(1, 100):
                points = pow_int(2, T - 1)
                if score2 + points >= 2 * N:
                    break

                prob = (
                    dp[score1 + 1][score2 + points]
                    + dp[score1][score2 + points]
                    + (pow_int(2, T) - 1) * dp[score1 + 1][score2]
                ) / (pow_int(2, T) + 1)

                if prob > dp[score1][score2]:
                    dp[score1][score2] = prob

    # Player 1's initial turn
    ans = (dp[0][0] + dp[1][0]) / 2
    return ans


def main() -> float:
    """Main entry point."""
    result = solve()
    print(f"{result:.8f}")
    return result


if __name__ == "__main__":
    main()

File: c/232.c
/*
 * Project Euler Problem 232: The Race
 *
 * Two players race to 100. Player 1 advances 1 on heads.
 * Player 2 chooses T>=1, flips T coins; advances 2^(T-1) if all heads (prob 1/2^T).
 * Find probability player 2 wins with optimal play.
 */
#include <stdio.h>

#define N 100

static double dp[N + 1][2 * N];

int main(void) {
    /* Base cases: player 2 wins if score2 >= N */
    for (int s1 = 0; s1 <= N; s1++)
        for (int s2 = N; s2 < 2 * N; s2++)
            dp[s1][s2] = 1.0;

    /* Fill DP backwards */
    for (int s1 = N - 1; s1 >= 0; s1--) {
        for (int s2 = N - 1; s2 >= 0; s2--) {
            double best = 0.0;
            for (int T = 1; T < 100; T++) {
                long long points = 1LL << (T - 1);
                if (s2 + points >= 2 * N) break;

                long long pow2T = 1LL << T;
                double prob = (dp[s1 + 1][s2 + (int)points]
                             + dp[s1][(int)(s2 + points)]
                             + (double)(pow2T - 1) * dp[s1 + 1][s2])
                             / (double)(pow2T + 1);

                if (prob > best) best = prob;
            }
            dp[s1][s2] = best;
        }
    }

    double ans = (dp[0][0] + dp[1][0]) / 2.0;
    printf("%.8f\n", ans);
    return 0;
}
